File: Algorithms/1_Recursion/javascript/reverse-string.js
//Implement a function that reverses a string using iteration and a function that uses recursion

function reverseStringIterative(str) {
  //O(n) - time complexity
  //O(n) space complexity -> result array is proportional to the input
  let reversed = [];
  let input = [...str];
  for (let i = 0; i < str.length; i++) {
    reversed.push(input.pop()); //[input.length - 1 - i]);
  }
  return reversed.join("");
}

function reverseStringIterative2(str) {
  //O(n) - time complexity
  //O(n) space complexity -> reversed string is proportional to the input
  let reversed = "";
  for (let i = str.length - 1; i >= 0; i--) {
    reversed += str[i];
  }
  return reversed;
}

//reverse string in place could help to achieve O(1) space complexity - BUT it is not done here because let strArray = str.split(""); is O(n) so it makes function O(n);
//JS strings are immutable
function reverseStringIterative3(str) {
  //O(n) - time complexity
  // Convert the string to an array since strings are immutable
  let strArray = str.split("");

  // Use two pointers approach to swap characters from the beginning and end
  let start = 0;
  let end = strArray.length - 1;

  while (start < end) {
    //swap characters at start/end indices
    let temp = strArray[start];
    strArray[start] = strArray[end];
    strArray[end] = temp;

    //move pointers towards each other
    end--;
    start++;
  }
  return strArray.join("");
}

function reverseStringRecursive(str) {
  // O(n)
  // Base case: if the string is empty or has only one character, it is already reversed
  if (str.length <= 1) {
    return str;
  }
  return reverseStringRecursive(str.substring(1)) + str.charAt(0);
  /*reverseStringRecursive("hello")
   reverseStringRecursive("ello") + "h"
      reverseStringRecursive("llo") + "e"
         reverseStringRecursive("lo") + "l"
            reverseStringRecursive("o") + "l"
               "o"
            Backtrack: "o" + "l" = "ol"
         Backtrack: "ol" + "l" = "oll"
      Backtrack: "oll" + "e" = "olle"
   Backtrack: "olle" + "h" = "olleh"
*/
}

console.log(reverseStringIterative("doog si azzip"));
console.log(reverseStringIterative2("doog si azzip"));
console.log(reverseStringIterative3("doog si azzip"));
console.log(reverseStringRecursive("doog si azzip"));
//should return: 'pizza is good'

File: Algorithms/1_Recursion/csharp/reverse-string-app/reverse-string.cs
using System;

class Program
{
  static string ReverseStringIterative(string str)
  {
    char[] reversed = new char[str.Length];
    char[] input = str.ToCharArray();

    for (int i = 0; i < str.Length; i++)
    {
      reversed[i] = input[^(i + 1)]; // ^ is the index from the end
    }

    return new string(reversed);
  }

  static string ReverseStringIterative2(string str)
  {
    char[] reversed = new char[str.Length];

    for (int i = str.Length - 1, j = 0; i >= 0; i--, j++)
    {
      reversed[j] = str[i];
    }

    return new string(reversed);
  }

  static string ReverseStringIterative3(string str)
  {
    char[] strArray = str.ToCharArray(); // O(n) space complexity due to the creation of char array -  O(1) space complexity if the input is char[] (in-place swapping)
    int start = 0;
    int end = strArray.Length - 1;

    while (start < end)
    {
      // Swap characters at start/end indices in place
      // char temp = strArray[start];
      // strArray[start] = strArray[end];
      // strArray[end] = temp;
      (strArray[start], strArray[end]) = (strArray[end], strArray[start]);

      // Move pointers towards each other
      end--;
      start++;
    }

    return new string(strArray);
  }

  static string ReverseStringIterative4(string str)
  {
    char[] stringArray = str.ToCharArray();
    Array.Reverse(stringArray);
    return new string(stringArray);
  }

  public static string ReverseStringRecursive(string str)
  {
    if (str.Length <= 1)
    {
      return str;
    }

    return ReverseStringRecursive(str[1..]) + str[0];
  }

  static void Main()
  {
    Console.WriteLine(ReverseStringIterative("doog si azzip"));
    Console.WriteLine(ReverseStringIterative2("doog si azzip"));
    Console.WriteLine(ReverseStringIterative3("doog si azzip"));
    Console.WriteLine(ReverseStringIterative4("doog si azzip"));
    Console.WriteLine(ReverseStringRecursive("doog si azzip"));
  }
}
